### PR TITLE
Fix Issue 19181: always error when an UFCS or opDispatch call has syntactically invalid arguments

### PR DESCRIPTION
**Pull request by FeepingCreature**
Repro:

```
$ dmd test.d
---
module test;

struct S {
    void opDispatch(string name, T)(T arg) { }
}

void main() {
    S s;
    s.foo(LanguageError);
}
```

Before:
    test.d(9): Error: no property foo for type S

After:
    test.d(9): Error: undefined identifier LanguageError

### CODE REVIEW

**Review by jacob-carlborg — APPROVED** [review]

**Comment by FeepingCreature** [comment]
Holy hell, that's the fastest approval I have *ever* gotten on Github, let alone DMD.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]
Needs to have a test.

**Comment by JinShil** [comment]
Doesn't this need a test case to prevent a regression.

**Comment by jacob-carlborg** [comment]
> Holy hell, that's the fastest approval I have ever gotten on Github, let alone DMD.

Sorry, I had to change my mind.

**Comment by jacob-carlborg** [comment]
> Doesn't this need a test case to prevent a regression.

Yes, I changed my mind and requested changes.

**Comment by FeepingCreature** [comment]
The original behavior is not strictly speaking _wrong_, it's just highly misleading. opDispatch fails to match since the call fails to compile - it just fails to compile due to a call-site error. That said, I'm not sure how you test for a specific error message.

**Comment by JinShil** [comment]
> That said, I'm not sure how you test for a specific error message.

See examples in the `test/fail_compilation` directory for examples.  Also see `TEST_OUTPUT` described in the Readme at https://github.com/dlang/dmd/tree/master/test

**Comment by FeepingCreature** [comment]
Test added.

File: test/fail_compilation/fail19181.d [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/fail19181.d(13): Error: undefined identifier LanguageError
+---
+*/
+struct S {
```
Comment (jacob-carlborg) [inline test/fail_compilation/fail19181.d]: Please follow the standard D style, curly braces on their own lines.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: test/fail_compilation/fail19181.d (hunk 1 above)
Comment (FeepingCreature) [reply test/fail_compilation/fail19181.d]: fixed

**Review by jacob-carlborg — APPROVED** [review]

File: test/fail_compilation/fail19181.d [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/fail19181.d(15): Error: undefined identifier LanguageError
```
Comment (JinShil) [inline test/fail_compilation/fail19181.d]: I think the tests suite will probably fail because LanguageError needs to be surrounded with tick marks.  We'll find out soon enough.

Comment (FeepingCreature) [reply test/fail_compilation/fail19181.d]: Since the other tests have that too, I've just gone ahead and added it. Can't hurt.

**Comment by jacob-carlborg** [comment]
Back to approved, still less than 20 minutes.

**Comment by FeepingCreature** [comment]
Still the fastest, I believe!

**Comment by FeepingCreature** [comment]
Again, that is not an error in my change.

**Comment by JinShil** [comment]
> Again, that is not an error in my change.

Are you sure?  Run the test suite locally with `make -C test -f Makefile`.  Or to run just the failing test use `rdmd test/run.d test/fail_compilation/ice14929.d`.

File: test/fail_compilation/ice14929.d [hunk 3]
```diff
@@ -75,7 +75,7 @@ template isComponentStorage(CS, C)
     {
         CS cs = CS.init;
         ulong eid;
-        cs.add(eid, c);
+        cs.add(eid, C.init);
```
Comment (jacob-carlborg) [inline test/fail_compilation/ice14929.d]: Why was this change necessary?

Comment (FeepingCreature) [reply test/fail_compilation/ice14929.d]: Fun reason: Previously, DMD would try to semantically evaluate `cs.add` _before_ figuring out what if anything `c` is, hit the `!in` instantiation error, and bail with the `TEST_OUTPUT` message. WIth this change, it verifies the arguments to cs.add() _first_ and immediately notices that there is no such variable as `c`, causing the `typeof` to fail and, as is `typeof`'s wont, silently return false.

Comment (jacob-carlborg) [reply test/fail_compilation/ice14929.d]: Ah, yes, of course. I was reading the corrected code and tried to figure out what was wrong with it :smiley:.

Comment (FeepingCreature) [reply test/fail_compilation/ice14929.d]: It is a bit weird and inconsistent that `typeof` errors out if only the error that happens in `typeof` is in a different template.

But that's neither what this bug nor the testcase's bug is about, so ... :man_shrugging:

**Review by wilzbach — APPROVED** [review]

**Comment by wilzbach** [comment]
> Or to run just the failing test use rdmd test/run.d test/fail_compilation/ice14929.d.

FYI:  for the lazy people like me: `./test/run.d fail_compilation/ice14929.d` and a few other combinations work too.

**Comment by MoonlightSentinel** [comment]
This introduced a regression: https://issues.dlang.org/show_bug.cgi?id=20488

**Comment by FeepingCreature** [comment]
(Regression fixed.)
